src/parser/core/Module.ts

```diff
@@ -0,0 +1,273 @@
+	 * @param {Error} error The error that occurred
+	 * @param {Object} event The event that was being processed when the error occurred, if source is 'event'
+	 * @returns {Object|undefined} The data to attach to automatic error reports, or undefined to rely on primitive value detection
+	 */
+	getErrorContext(source: 'event' | 'output', error: Error, event?: Event): any {
+		return
+	}
+
+	protected addHook<T extends Event>(
+		events: T['type'] | T['type'][],
+		cb: HookCallback<T>
+	): Hook<T>
+	protected addHook<T extends Event>(
+		events: T['type'] | T['type'][],
+		filter: Filter<T> | HookCallback<T>,
+		cb: HookCallback<T>,
+	): Hook<T>
+	protected addHook<T extends Event>(
+		events: T['type'] | T['type'][],
+		_filter: Filter<T> | HookCallback<T>,
+		_cb?: HookCallback<T>,
+	): Hook<T> | undefined {
+		// I'm currently handling hooks at the module level
+		// Should performance become a concern, this can be moved up to the Parser without breaking the API
+		const cb = typeof _filter === 'function'? _filter : _cb
+		let filter = typeof _filter === 'function'? {} : _filter
+
+		// If there's no callback just... stop
+		if (!cb) { return }
+
+		// QoL filter transforms
+		filter = this.mapFilterEntity(filter, 'to', 'targetID')
+		filter = this.mapFilterEntity(filter, 'by', 'sourceID')
+		if (filter.abilityId) {
+			if (!filter.ability) {
+				filter.ability = {}
+			}
+			filter.ability.guid = filter.abilityId
+			delete filter.abilityId
+		}
+
+		// Make sure events is an array
+		if (!Array.isArray(events)) {
+			events = [events]
+		}
+
+		// Final hook representation
+		const hook = {
+			events,
+			filter,
+			callback: cb.bind(this),
+		}
+
+		// Hook for each of the events
+		events.forEach(event => {
+			let hooks = this._hooks.get(event)
+
+			// Make sure the map has a key for us
+			if (!hooks) {
+				hooks = new Set()
+				this._hooks.set(event, hooks)
+			}
+
+			// Set the hook
+			hooks.add(hook)
+		})
+
+		// Return the hook representation so it can be removed (later)
+		return hook
+	}
+
+	private mapFilterEntity<T extends Event>(
+		filter: Filter<T>,
+		qol: keyof Filter<T>,
+		raw: keyof T,
+	) {
+		if (!filter[qol]) { return filter }
+
+		const _filter = cloneDeep(filter)
+
+		// TODO: Typing on parser req. for some of this stuff
+		switch (_filter[qol]) {
+			case 'player':
+				_filter[raw] = this.parser.player.id
+				break
+			case 'pet':
+				_filter[raw] = this.parser.player.pets.map((pet: any) => pet.id)
+				break
+			default:
+				_filter[raw] = _filter[qol]
+		}
+
+		delete _filter[qol]
+
+		return _filter
+	}
+
+	// TODO: Test
+	protected removeHook(hook: Hook<any>) {
+		hook.events.forEach(event => {
+			const hooks = this._hooks.get(event)
+			if (!hooks) { return }
+			hooks.delete(hook)
+		})
+	}
+
+	triggerEvent(event: Event) {
+		// Run through registered hooks. Avoid calling 'all' on symbols, they're internal stuff.
+		if (typeof event.type !== 'symbol') {
```

Comment:
If you just believe the type definitions this looks like unreachable code; better change the definition of `type` to be `string | symbol`.

By the way, if the symbol is created with `Symbol()`, you can take its `typeof` to make a `unique symbol`: `const foo = Symbol(); type Foo = typeof foo`. The `typeof` of a symbol is pretty much the only type that's arguably nominal in typescript (no `unique symbol` is equal to another `unique symbol`).